Proposal for improving vulnerability detection MacOS

juan-fdz-hawa opened the pull request:
Relates to #6001 

Proposal for improving vulnerability detection on MacOS

chiiph reviewed (dismissed):
This sounds good to me. We would still have the gap of the CVEs that don't have CPE, correct?

We can start by mapping the top 20 apps properly and then evolve from there.

On the other hand, I wonder... could we create our own OVAL files for macOS and Windows based on the NVD database and just use OVAL everywhere? Then we can focus on improving the OVAL definition instead of the code.

juan-fdz-hawa commented:
> This sounds good to me. We would still have the gap of the CVEs that don't have CPE, correct?
> 
> We can start by mapping the top 20 apps properly and then evolve from there.
> 
> On the other hand, I wonder... could we create our own OVAL files for macOS and Windows based on the NVD database and just use OVAL everywhere? Then we can focus on improving the OVAL definition instead of the code.

Yup, this proposal only addresses the issues with have with finding the proper entries in the NVD dataset - we still have the issue that for some software classes, the NVD dataset seems to be out of date.

For testing this, I installed most of the apps listed here: https://github.com/fleetdm/fleet/issues/5691 (see the `apps` sheet) and then made sure that if an app was 'not found' in the NVD dataset was because there were no NVD entries for that app (that's how I came up with the translation tables). So everything in the `not_found` sheet is because there are no NVD entries for those items.

The biggest disadvantage of this approach is that we will need translation tables. I tried using approximations based on text distance (both Levenshtein distance and the gestalt pattern matching algo) - but at least for MacOS it was not worth it  (because when using approximations there's always the probability that we will get it wrong, and then we need to show that somehow in the UI and maybe get user input on that) - for Windows the story might be different because the set of available software is much bigger, and then having translation tables becomes a PITA.

Re creating OVAL file: We could, but we will still need to solve software the binding problem, also (and this might not be a problem) both OVAL and the NVD approach have different 'semantics' - the OVAL approach needs to analyze the host as a whole, whereas the NVD approach only looks at software classes.

chiiph commented:
> Re creating OVAL file: We could, but we will still need to solve software the binding problem, also (and this might not be a problem) both OVAL and the NVD approach have different 'semantics' - the OVAL approach needs to analyze the host as a whole, whereas the NVD approach only looks at software classes.

Yeah, the same problems remain, but maybe we can have only one vulnerability detection approach with this... just a random thought though, please disregard if it doesn't sound like it has a significant upside.

juan-fdz-hawa commented:
> > Re creating OVAL file: We could, but we will still need to solve software the binding problem, also (and this might not be a problem) both OVAL and the NVD approach have different 'semantics' - the OVAL approach needs to analyze the host as a whole, whereas the NVD approach only looks at software classes.
> 
> Yeah, the same problems remain, but maybe we can have only one vulnerability detection approach with this... just a random thought though, please disregard if it doesn't sound like it has a significant upside.

No, I think it was a nice idea! (sorry if my previous comment come out a little dismissive, that was not the intention) - the way things are looking, I think if we want really good vulnerability detection we will need to consume various heterogeneous data sources, so having some kind of 'intermediate' format that we can just feed to some 'rule-based' engine will simplify the architecture a lot - so definitely- something to keep in mind

lucasmrod commented on proposals/improv-mac-os-vuln-detection.md:
```diff
@@ -0,0 +1,117 @@
+current approach to vulnerability detection on MacOS:
+
+- The version reported by software does not fit the standard format. For example, Zoom reports the version as 5.8.3 (2240).
+- The app name includes extra terms that don't appear in the title. For example, zoom.us is treated
```
Am a bit out of the loop, could you define the following two?
- `app name`? (is this is the app name as defined by osquery?)
- `title`? (does this refer to the title in the `cpe` database?)

juan-fdz-hawa commented on proposals/improv-mac-os-vuln-detection.md:
Sorry, didn't include anything related to that on this proposal - the approach I would follow is to see if Apple has any kind of standard regarding on how versions  are specified and use that to try and parse the reported version. If not, like you mentioned, we would need to normalize the version somehow.
